lib/wal/wal.c

```diff
@@ -113,46 +174,50 @@ wal_timer(struct work_struct *work)
 
         rid = atomic_read(&wal->wal_rid);
         if (rid != rid_last || closing) {
+            struct wal_flush_stats stats;
             rid_last = rid;
 
-            err = wal_bufset_flush(wal->wbs, &flushb, &bufsz, &buflen);
+            err = wal_bufset_flush(wal->wbs, &stats);
             if (err) {
                 atomic_set(&wal->error, err);
                 wal_ionotify_cb(wal, err); /* Notify sync waiters on flush error */
                 continue;
             }
 
-            /* No dirty data, notify any sync waiters */
-            if (flushb == 0)
-                wal_ionotify_cb(wal, 0);
-
-            if (wal->wal_thr_sensor) {
-                const uint64_t hwm = (bufsz * wal->wal_thr_hwm) / 100;
-                const uint64_t lwm = (bufsz * wal->wal_thr_lwm) / 100;
-                uint32_t new;
+            if (stats.flush_tlen == 0)
+                wal_ionotify_cb(wal, 0); /* No dirty data, notify any sync waiters */
 
-                ev(buflen >= bufsz);
+            wal_throttle_sensor_set(wal, stats.bufsz, stats.max_buflen);
 
-                new = (buflen > lwm) ? (THROTTLE_SENSOR_SCALE * buflen) / hwm : 0;
-
-                throttle_sensor_set(wal->wal_thr_sensor, new);
-            }
+            wal_flush_wait(wal, &stats, WAL_FLUSH_WAIT_PCT);
 
             lag = get_time_ns() - tstart;
             sleep_ns = (lag >= sleep_ns || closing) ? 0 : sleep_ns - lag;
         } else {
-            /* No mutations, notify any sync waiters */
-            wal_ionotify_cb(wal, 0);
+            wal_ionotify_cb(wal, 0); /* No mutations, notify any sync waiters */
         }
 
         mutex_lock(&wal->timer_mutex);
         end_stats_work();
 
-        if (wal->sync_pending)
+        if (wal->sync_pending) {
             closing = false;
-        else if (!closing && sleep_ns > 0)
-            cv_timedwait(&wal->timer_cv, &wal->timer_mutex,
-                         NSEC_TO_MSEC(sleep_ns), "waltmslp");
+        } else if (!closing && sleep_ns > 0) {
+            uint64_t flushv[WAL_BUF_MAX], intvl, tstart = get_time_ns();
+            int flushc;
+
+            intvl = max_t(uint64_t, sleep_ns / 10, MSEC_TO_NSEC(1));
+            flushc = wal_bufset_flushoff(wal->wbs, WAL_BUF_MAX, flushv);
+
+            while (cv_timedwait(&wal->timer_cv, &wal->timer_mutex,
+                                NSEC_TO_MSEC(intvl), "waltmslp") == ETIMEDOUT &&
+                   !wal->sync_pending && (get_time_ns() - tstart < sleep_ns) &&
+                   !atomic_read(&wal->closing)) {
+                if (wal_dirty_exceeds_threshold(wal, flushc, flushv))
+                    break;
+            }
```

Comment:
Seems like before the throttle would adjust slowly as the buffers fill, but now maybe it just suddenly jumps when we exit this new inner loop?

Reply:
Yes, it will be updated more frequently than what it used to.. I didn't find issues due to that in my tests - for instance, my ycsb native load with this patch went up ~5%..  Maybe updating it frequently helps to avoid sudden jumps in the sensor values?

Reply:
Another thing that could help us here is the uniform IO load over time rather than bursts every 100ms..